Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,212 @@
+@@ -360,9 +366,9 @@ public abstract class ChunkMapDistance {
+         }
+ 
+         @Override
+-        protected void a(long i, int j, int k) {
+-            this.g.add(i);
+-        }
++        protected void a(long i, int j, int k, byte d) {
++            this.insertUpdate(i, d);
++        } //Paper ticket level pp-view-dist
+ 
+         public void a(int i) {
+             ObjectIterator objectiterator = this.a.long2ByteEntrySet().iterator();
+@@ -372,20 +378,20 @@ public abstract class ChunkMapDistance {
+                 byte b0 = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getByteValue();
+                 long j = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getLongKey();
+ 
+-                this.a(j, b0, this.c(b0), b0 <= i - 2);
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i); //Paper ticket level pp-view-dist
+             }
+ 
+             this.e = i;
+         }
+ 
+-        private void a(long i, int j, boolean flag, boolean flag1) {
+-            if (flag != flag1) {
++        private void a(long i, int j, boolean flag, boolean flag1, int vd) { //Paper ticket level pp-view-dist
++            if (flag != flag1 || (flag && flag1)) { //Paper ticket level pp-view-dist
+                 Ticket<?> ticket = new Ticket<>(TicketType.PLAYER, ChunkMapDistance.b, new ChunkCoordIntPair(i), ChunkMapDistance.this.currentTick);
+ 
+                 if (flag1) {
+                     ChunkMapDistance.this.j.a(ChunkTaskQueueSorter.a(() -> { // Craftbukkit - decompile error
+                         ChunkMapDistance.this.m.execute(() -> {
+-                            if (this.c(this.c(i))) {
++                            if (this.getChunksMapping(i) <= vd - 2) { //Paper ticket level pp-view-dist
+                                 ChunkMapDistance.this.addTicket(i, ticket);
+                                 ChunkMapDistance.this.l.add(i);
+                             } else {
+@@ -412,10 +418,14 @@ public abstract class ChunkMapDistance {
+         public void a() {
+             super.a();
+             if (!this.g.isEmpty()) {
+-                LongIterator longiterator = this.g.iterator();
++                //Paper start ticket level pp-view-dist
++                ObjectIterator<Long2ByteMap.Entry> longiterator = this.getUpdateIterator();
+ 
+                 while (longiterator.hasNext()) {
+-                    long i = longiterator.nextLong();
++                    Long2ByteMap.Entry e = longiterator.next();
++                    long i = e.getLongKey();
++                    int vd = e.getByteValue();
++                    //Paper end
+                     int j = this.f.get(i);
+                     int k = this.c(i);
+ 
+@@ -430,7 +440,7 @@ public abstract class ChunkMapDistance {
+                             }
+ 
+                         });
+-                        this.a(i, k, this.c(j), this.c(k));
++                        this.a(i, k, this.c(j), k <= vd - 2, vd); //Paper ticket level pp-view-dist
```

Comment:
still checking against global view distance (the check range call)

Reply:
this is intentional.

Reply:
you are checking against the wrong view distance, why